File: Solutions/Maximum_Product_Subarray/Maximum_Product_Subarray.py
[Python] dp solution, explained

#Let dp1[i] be the maximum product which ends with i-th index and dp2[i] be the minimum product which ends with i-the index. Why we need both maximum and minimum? Because we can have big negative number, which then multiplied by negative will give big positive number.
#
#So, let us traverse all numbers and update our dp1 and dp2:

#1.If nums[k] > 0, then biggest number dp1[k] can be obtained from dp1[k-1] * nums[k] or nums[k]. The smallest number dp2[k] is equal to dp2[k-1] * nums[k].
#2.If nums[k] <= 0, then the biggest number is dp1[k] = dp2[k-1] * nums[k] and the smallest number is dp2[k] = min(dp1[k-1] * nums[k], nums[k]).
#3.Finally, we return maximum of dp1.
#4Complexity: both time and space is O(n). Space complexity can be improved to O(1), because we always use only previous elements in our dp1 and dp2 tables.

class Solution:
    def maxProduct(self, nums):
        N = len(nums)
        dp1 = [0] * N
        dp2 = [0] * N
        dp1[0] = dp2[0] = nums[0]
        
        for k in range(1, N):
            if nums[k] > 0:
                dp1[k] = max(dp1[k-1] * nums[k], nums[k])
                dp2[k] = dp2[k-1] * nums[k]
            else:
                dp1[k] = dp2[k-1] * nums[k]
                dp2[k] = min(dp1[k-1] * nums[k], nums[k])
                
        return max(dp1)

#Fastest
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if (len(nums) == 1):
            return nums[0]
        if 0 in nums:
            nums = self.splitByZero(nums)
            products = []
            
            for arr in nums:
                products.append(self.maxProductNoZeros(arr))
                
            products.append(0)
            print(products)
            return max(products)
        else:
            return self.maxProductNoZeros(nums)
            
    def splitByZero(self, nums):
        retArr = [] #2d array
        row = []
        for num in nums:
            if num == 0:
                if len(row) > 0:
                    retArr.append(row)
                    row = []
            else:
                row.append(num)
        if len(row) > 0:
            retArr.append(row)
        return retArr
    
    def maxProductNoZeros(self, nums):        
        if (len(nums) == 1):
            return nums[0]
        
        countNeg = 0
        for num in nums:
            if num < 0:
                countNeg += 1
                
        if (countNeg % 2) == 0:
            #include whole array
            prod = 1
            for num in nums:
                prod *= num
            return prod
        
        #odd number of negatives in array - find whether we should eliminate the last or the first
        elimFirstProd = 1
        i = 0
        while nums[i] > 0 and i < len(nums):
            i += 1
        while i < len(nums)-1:
            i += 1
            elimFirstProd *= nums[i]
            
            
        elimLastProd = 1
        i = len(nums)-1
        while (nums[i] > 0) and i >= 0:
            i -= 1
        while i > 0:
            i -= 1
            elimLastProd *= nums[i]
            
            
        return max(elimFirstProd, elimLastProd)

File: Solutions/Maximum_Product_Subarray/Maximum_Product_Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n =  nums.size();
        vector<int> dp1(n, 0), dp2(n, 0);
        dp1[0] = nums[0];
        dp2[0] = nums[0];
        int result = nums[0];
        
        for(int i=1; i<n; i++) {
            if(nums[i] > 0) {
                dp1[i] = max(dp1[i-1] * nums[i], nums[i]);
                dp2[i] = dp2[i-1] * nums[i];
            } else {
                dp1[i] = dp2[i-1] * nums[i];
                dp2[i] = min(dp1[i-1] * nums[i], nums[i]);
            }
            result = max(result, dp1[i]);
        }
        
        return result;
    }
};

//Better
class Solution {
public:
    int maxProduct(vector<int>& nums) 
    {
		// empty array case
		if (nums.size() == 0)
			return 0;
			
	    // maxSub and minSub will hold the products till nums[i]
        int maxSub = nums[0];   
        int minSub = nums[0];
        int maxProductSub = nums[0];
        
        for (size_t i = 1; i < nums.size(); i++)
        {
            // element is negative so we swap max and min
            // because when multiplying negative with a negative, number becomes positive so minimum negative number will become the maximum number
            if (nums[i] < 0)
                swap(minSub, maxSub);
      
            // update all the sub values
			maxSub = max(maxSub * nums[i], nums[i]); 
            minSub = min(minSub * nums[i], nums[i]); 
            // choose max product to be the max between the maxProduct till now and maxSub
			maxProductSub = max(maxProductSub, maxSub); 
        }
		
        return maxProductSub;
    }
   
};

//Fastest
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res=nums[0], ma=nums[0], mi=nums[0];
        for (int i=1; i<nums.size(); i++) {
            if (nums[i]<0)
                swap(ma, mi);
            
            ma = max(ma*nums[i], nums[i]);
            mi = min(mi*nums[i], nums[i]);
            res = max(res, ma);
        }
        return res;
    }
};